From 8e0b82b863822255030db60df7fe2fcbe940bb09 Mon Sep 17 00:00:00 2001
From: Daire McNamara <daire.mcnamara@microchip.com>
Date: Mon, 13 Dec 2021 17:08:55 +0000
Subject: [PATCH] Remap DTB explicitly

---
 arch/riscv/mm/init.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/arch/riscv/mm/init.c b/arch/riscv/mm/init.c
index 067583ab1bd7..216990621822 100644
--- a/arch/riscv/mm/init.c
+++ b/arch/riscv/mm/init.c
@@ -512,6 +512,14 @@ static void __init setup_vm_final(void)
 			   __pa_symbol(fixmap_pgd_next),
 			   PGDIR_SIZE, PAGE_TABLE);
 
+#ifndef CONFIG_BUILTIN_DTB
+	/* Remap DTB explicitly; make no assumptions about its location */
+	pa = dtb_early_pa & ~(PMD_SIZE - 1);
+	va = (uintptr_t)__va(pa);
+	create_pgd_mapping(swapper_pg_dir, va, pa, PMD_SIZE, PAGE_KERNEL);
+	create_pgd_mapping(swapper_pg_dir, va + PMD_SIZE, pa + PMD_SIZE, PMD_SIZE, PAGE_KERNEL);
+#endif
+
 	/* Map all memory banks */
 	for_each_mem_range(i, &start, &end) {
 		if (start >= end)
